### PR TITLE
drivers: defined battery driver api's.

### PR DESCRIPTION
**Pull request by rkotabag**
Defined API's header file for new battery driver.

These API defined for TI Fuel Gauge based SBS battery. These API are defined as per the current application usage. 
Please note that we have defined battery and charger as separate drivers.  

Signed-off-by: Ramesh Kotabagi <ramesh.kotabagi@intel.com>

### CODE REVIEW

**Review by pabigot — CHANGES_REQUESTED** [review]
It would be OK to add this as a specific driver for a specific battery management capability, but a generic API needs to be developed taking into account #2542.

There may also be some precedent for treating this functionality as a sensor, e.g. `ti,bq274xx`.

File: drivers/Kconfig [hunk 1]
```diff
@@ -97,4 +97,6 @@ source "drivers/eeprom/Kconfig"
 
 source "drivers/peci/Kconfig"
 
+source "drivers/battery/Kconfig"
```
Comment (nashif) [inline drivers/Kconfig:100]: driver infrastructure and Kconfig should come with a driver.

File: drivers/battery/Kconfig [hunk 2]
```diff
@@ -0,0 +1,24 @@
+# Battery configuration options
```
Comment (nashif) [inline drivers/battery/Kconfig:1]: we should not add Kconfigs without actually having a user

File: dts/bindings/battery/intel,sbs_battery.yaml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) 2020 Intel Corporation
```
Comment (nashif) [inline dts/bindings/battery/intel,sbs_battery.yaml:1]: this should come with the driver, not part of this PR introducing an API

File: include/drivers/battery.h [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Battery public API header file.
+ */
+
+#ifndef __BATTERY_H__
```
Comment (nashif) [inline include/drivers/battery.h:12]: follow guidelines, should be ZEPHYR_INCLUDE_DRIVERS_BATTERY_H_

**Review by nashif — CHANGES_REQUESTED** [review]

**Comment by nashif** [comment]
> It would be OK to add this as a specific driver for a specific battery management capability, but a generic API needs to be developed taking into account #2542.

please be specific with what you are asking for or summaries the relevant part of #2542.

File: drivers/Kconfig (hunk 1 above)
Comment (rgundi) [reply drivers/Kconfig:100]: This is implemented based on the discussions in https://github.com/zephyrproject-rtos/zephyr/issues/5554. The actual driver implementation is defined out of tree. Not sure how to open that up for review.

File: dts/bindings/battery/intel,sbs_battery.yaml (hunk 3 above)
Comment (rgundi) [reply dts/bindings/battery/intel,sbs_battery.yaml:1]: Yes. This is a question i have too. If we have to keep a driver implementation out of tree, how do we express it in dts?

**Comment by rkotabag** [comment]
> This looks very interesting. However, I don't fully understand the target of this driver. Looking at the code, it seems to be based on a specific I2C-connected chip and not targeting a generic battery API. Can you please give a bit more background?



> This looks very interesting. However, I don't fully understand the target of this driver. Looking at the code, it seems to be based on a specific I2C-connected chip and not targeting a generic battery API. Can you please give a bit more background?

File: dts/bindings/battery/intel,sbs_battery.yaml (hunk 3 above)
Comment (nashif) [reply dts/bindings/battery/intel,sbs_battery.yaml:1]: You can also have DTS data out of tree.

**Comment by rkotabag** [comment]
> This looks very interesting. However, I don't fully understand the target of this driver. Looking at the code, it seems to be based on a specific I2C-connected chip and not targeting a generic battery API. Can you please give a bit more background?

These battery API's defined for TI Fuel Gauge based SBS battery. These API's are defined as per the current application usage. 
Please note that we have defined battery and charger as separate drivers.

**Comment by pabigot** [comment]
#5554 is particularly relevant to drivers for existing subsystem APIs like gpio or sensor, where the specific driver implementation is not being upstreamed.

This is different, because there is no existing subsystem API.  Defining what it might look like is what #2542 is about.

Drivers don't have to be based on a subsystem abstraction.  It seems likely both the drivers that motivate the proposed APIs exist already out of tree; if not, they certainly could be implemented without an abstraction, or with an out-of-tree abstraction.  So if the drivers are expected to remain out of tree I'd suggest just doing that.

If you want drivers for these systems in upstream Zephyr then your fastest path forward is to do them without an abstraction.  Then they could be used as a basis for identifying what a generic API should look like by comparing them with other specific drivers to identify the common capabilities.

It's possible that you could get an API for a charger subsystem accepted if the drivers that implement the API were part of the PR, but I'm not the only one who insists that any new API be validated by the presence of at least two accompanying implementations from different vendors that show the API provides the desired functionality and isn't too tightly bound to the details of a specific driver.

As I noted there is an existing TI Fuel Gauge driver in-tree that is exposed as a sensor.  That also would be a path to follow that's more likely to be successful than creating a new API.

If you do want to create a new API:
* It should take into account the requirements discussed in #2542
* The work should include a discussion of how the TI BQ27421 could be converted to the API
* It should combine the battery and charger APIs or explain why they should be kept separate
* It will likely be delayed until additional drivers have been written to validate the API https://github.com/zephyrproject-rtos/zephyr/issues/2542#issuecomment-606706804

**Comment by nashif** [comment]
>  but I'm not the only one who insists that any new API be validated by the presence of at least two accompanying implementations from different vendors that show the API provides the desired functionality and isn't too tightly bound to the details of a specific driver.

This is not a requirement for introducing a new API, this is a requirement for making an API stable. We need to start somewhere and we need at least one implementation to get an API into an experimental state.

See https://docs.zephyrproject.org/latest/development_process/api_lifecycle.html#experimental

https://github.com/zephyrproject-rtos/zephyr/issues/2542#issuecomment-606706804 is not a zephyr requirement. A driver can be added with an API and considered experimental, more drivers implementing this API would then move it to a different state after 2 releases.

**Comment by nashif** [comment]
btw, #2542 is about energy management and hardware independent APIs, this here is about support for peripherals and device drivers that would feed into such framework. 

> It should take into account the requirements discussed in #2542

Would be great if you could list those here as specific requirements, it is not clear to me which requirements are you referring to.

**Comment by carlescufi** [comment]
@rkotabag are you planning to add an implementation of this driver? This is a requirement for any new API:

>The API introduction should be accompanied by at least one implementation of said API (in the case of peripheral APIs, this corresponds to one driver)

https://docs.zephyrproject.org/latest/development_process/api_lifecycle.html#experimental

**Comment by rkotabag** [comment]
> @rkotabag are you planning to add an implementation of this driver? This is a requirement for any new API:
> 
> > The API introduction should be accompanied by at least one implementation of said API (in the case of peripheral APIs, this corresponds to one driver)
> 
> https://docs.zephyrproject.org/latest/development_process/api_lifecycle.html#experimental

Up streaming the driver implementation for defined battery API's is currently dropped after discussing with our internal team, due to the following reasons,
1. Defined API and driver implementation are specific to particular chip and we are not sure w.r.t NDA and IP regarding same to open source the implementation.
2. Defined API's does not  meet 100% expectations of #2542 
So its decided internally that we will modify the current API for charger and battery driver to meet the requirement #2542  

So this pull request will be closed and will be working on to meet the requirement of #2542 .

**Comment by rkotabag** [comment]
Up streaming the driver implementation for defined battery API's is currently dropped after discussing with our internal team, due to the following reasons,
1. Defined API and driver implementation are specific to particular chip and we are not sure w.r.t NDA and IP regarding same to open source the implementation.
2.Defined API's does not meet 100% expectations of #2542

So its decided that we will be working on to meet the requirement of #2542  and close this pull request.

I am closing the this pull request as commented above. 
I thank everyone for there valuable inputs and exposure to #2542 requirement.
